### ISSUE TITLE
AWS IoT: AWSIotMqttManager can't trigger Just-in-Time Registration

### ISSUE THREAD

**Issue by pwa16000** [aws-amplify/aws-sdk-android#451]
### To help us solve your problem better, please answer the following list of questions.

* What service are you using?
AWS IoT

* In what version of SDK are you facing the problem?
com.amazonaws:aws-android-sdk-iot:2.6.18

* Is the issue limited to Simulators / Actual Devices?
No, both not as expected

* Can your problem be resolved if you bump to a higher version of SDK?
No, Latest SDK was used

* Is this problem related to specific Android/OS version?
No

* Can you give us steps to reproduce with a minimal, complete, and verifiable example? Please include any specific network conditions that might be required to reproduce the problem.
Just-in-Time Registration is one method to manage X.509, URL: https://aws.amazon.com/cn/blogs/iot/just-in-time-registration-of-device-certificates-on-aws-iot/ described detail steps on how to use this method.

I tried all steps in URL https://aws.amazon.com/cn/blogs/iot/just-in-time-registration-of-device-certificates-on-aws-iot/  and it can work. most import step in that URL is “When a device attempts to connect with an X.509 certificate that is not known to AWS IoT but was signed by a CA that was registered with AWS IoT, the device certificate will be auto-registered by AWS IoT in a new PENDING_ACTIVATION state.”，

If I following suggestion on that URL, use mosquitto_pub to send connect request to AWS IoT, it can trigger publishes a message to the following MQTT topic($aws/events/certificates/registered/<caCertificateID>).  but if I use AWSIotMqttManager to connect to AWS IoT, it can't trigger such publishes, so Just-in-Time Registration does not work with IoT aws-sdk-android.

I modify code under https://github.com/awslabs/aws-sdk-android-samples/tree/master/AndroidPubSub to test this scenario, updated java file as attached.
[PubSubActivity.zip](https://github.com/aws/aws-sdk-android/files/1946435/PubSubActivity.zip)

Why AWSIotMqttManager.connect can't trigger Just-in-Time Registration? 

Is there any way to bypass it?

Thanks

* Please include a stacktrace if applicable.

If you need help with understanding how to implement something in parqticular then we suggest that you first look into our [developer guide](https://docs.aws.amazon.com/mobile/sdkforandroid/developerguide/). You can also simplify your process of creating an application by using [Mobile Hub](https://console.aws.amazon.com/mobilehub/home#/).

**Comment by pwa16000** [comment]
I re-read ug of AWS iot at https://docs.aws.amazon.com/iot/latest/developerguide/device-certs-your-own.html.

It said: When a device first attempts to connect to AWS IoT, as part of the TLS handshake, it must present a registered CA certificate and a device certificate. AWS IoT recognizes the CA certificate as a registered CA certificate and automatically registers the device certificate and sets its status to PENDING_ACTIVATION. This means the device certificate was automatically registered and is awaiting activation. A certificate must be in the ACTIVE state before it can be used to connect to AWS IoT. When AWS IoT automatically registers a certificate or when a certificate in PENDING_ACTIVATION status connects, AWS IoT publishes a message to the following MQTT topic:

$aws/events/certificates/registered/caCertificateID

Looks this step does not happen when I use AWSIotMqttManager.connect with device certificate does not in active state.

**Comment by kvasukib** [comment]
The SDK uses Paho MQTT Client currently. The AWSIotMqttManager doesn't support JITR out of the box.

In order to register and activate your own certificate in the first connect attempt, you will need to send your CA certs along with the certificate. Hence, you need to add the CA Certs into the certificate chain. We'll add this support in the future release of the SDK. 

In the meantime, you could try to add that in the following lines: please refer to https://github.com/aws/aws-sdk-android/blob/master/aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotKeystoreHelper.java#L297

tempKeystore.setKeyEntry("key-alias", key,
AWS_IOT_INTERNAL_KEYSTORE_PASSWORD.toCharArray(), new Certificate[] {
cert, caCert });

**Comment by pwa16000** [comment]
Hi, @kvasukib, Thanks for your suggestion, I will try.

**Comment by muellerfr** [comment]
This issue has been closed because of inactivity. Please open a new issue if are still encountering problems.

**Comment by kvasukib** [comment]
@pwa16000 

Typically you will do the following if you are using a certificate issued and signed by AWS IoT.
You will create the certificate and then save the certificate and private key in the keystore and you pass the keystore instance in the `connect` method.

```java
        // create a new certificate and private key
        CreateKeysAndCertificateRequest certRequest = new CreateKeysAndCertificateRequest();
        certRequest.setSetAsActive(true);
        this.certResult = iotClient.createKeysAndCertificate(certRequest);

        // save certificate and private key in a keystore
        AWSIotKeystoreHelper.saveCertificateAndPrivateKey(this.certResult.getCertificateId(),
                certResult.getCertificatePem(),
                certResult.getKeyPair().getPrivateKey(),
                KEYSTORE_PATH,
                KEYSTORE_NAME,
                KEYSTORE_PASSWORD);

        // retrieve the keystore
        KeyStore ks = AWSIotKeystoreHelper.getIotKeystore(certResult.getCertificateId(),
                KEYSTORE_PATH,
                KEYSTORE_NAME,
                KEYSTORE_PASSWORD);

        // connect to AWS IoT using keystore
        mqttManager.connect(ks, new AWSIotMqttClientStatusCallback() {
            @Override
            public void onStatusChanged(AWSIotMqttClientStatus status, Throwable throwable) {
                statuses.add(status);
            }
        });
```

My suggestion as a workaround would be to concatenate the certificate with the CA certificate and store in the keystore and see if that unblocks the authentication to connect method.

@pwa16000 Do you have any results from your testing?

**Comment by jackie099** [comment]
Any update to this please?

**Comment by ilcannibale** [comment]
After one year i have the same error yet. I doesn't really understand why the caCert parameter was removed. I am really discouraged to use this sdk.

**Comment by gpanshu** [comment]
Is this still a problem ? We can do a deep dive and try to repro if it is.

**Comment by div5yesh** [comment]
This issue has been closed because of inactivity. Please open a new issue if are still encountering problems.
